File: backend/order-service/src/main/java/com/fooddelivery/orderservice/model/Order.java
package com.fooddelivery.orderservice.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "orders")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long customerId;

    @Column(nullable = false)
    private Long restaurantId;

    @Column(nullable = false)
    private Long deliveryAddressId; // Reference to customer's address

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<OrderItem> items = new ArrayList<>();

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal subtotal; // Sum of all items

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal deliveryFee;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal tax;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal total; // subtotal + deliveryFee + tax

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status = OrderStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentStatus paymentStatus = PaymentStatus.PENDING;

    @Column(length = 500)
    private String specialInstructions;

    private LocalDateTime estimatedDeliveryTime;

    private LocalDateTime actualDeliveryTime;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    // Helper method to calculate total
    public void calculateTotal() {
        this.subtotal = items.stream()
                .map(item -> item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        this.total = subtotal.add(deliveryFee).add(tax);
    }

    // Helper method to add item
    public void addItem(OrderItem item) {
        items.add(item);
        item.setOrder(this);
    }
}

File: frontend/lib/models/order.dart
class OrderItem {
  final int id;
  final int menuItemId;
  final String itemName;
  final int quantity;
  final double price;
  final String? customizations;

  OrderItem({
    required this.id,
    required this.menuItemId,
    required this.itemName,
    required this.quantity,
    required this.price,
    this.customizations,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'],
      menuItemId: json['menuItemId'],
      itemName: json['itemName'],
      quantity: json['quantity'],
      price: (json['price'] as num).toDouble(),
      customizations: json['customizations'],
    );
  }
}

class Order {
  final int id;
  final int customerId;
  final int restaurantId;
  final int deliveryAddressId;
  final List<OrderItem> items;

  final double subtotal;
  final double deliveryFee;
  final double tax;
  final double total;

  final String status;
  final String paymentStatus;
  final String? specialInstructions;

  final String? estimatedDeliveryTime;
  final String? actualDeliveryTime;
  final String? createdAt;
  final String? updatedAt;

  Order({
    required this.id,
    required this.customerId,
    required this.restaurantId,
    required this.deliveryAddressId,
    required this.items,
    required this.subtotal,
    required this.deliveryFee,
    required this.tax,
    required this.total,
    required this.status,
    required this.paymentStatus,
    this.specialInstructions,
    this.estimatedDeliveryTime,
    this.actualDeliveryTime,
    this.createdAt,
    this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    final itemsJson = (json['items'] as List?) ?? [];
    return Order(
      id: json['id'],
      customerId: json['customerId'],
      restaurantId: json['restaurantId'],
      deliveryAddressId: json['deliveryAddressId'],
      items: itemsJson.map((e) => OrderItem.fromJson(e)).toList().cast<OrderItem>(),
      subtotal: (json['subtotal'] as num).toDouble(),
      deliveryFee: (json['deliveryFee'] as num).toDouble(),
      tax: (json['tax'] as num).toDouble(),
      total: (json['total'] as num).toDouble(),
      status: json['status']?.toString() ?? "UNKNOWN",
      paymentStatus: json['paymentStatus']?.toString() ?? "UNKNOWN",
      specialInstructions: json['specialInstructions'],
      estimatedDeliveryTime: json['estimatedDeliveryTime']?.toString(),
      actualDeliveryTime: json['actualDeliveryTime']?.toString(),
      createdAt: json['createdAt']?.toString(),
      updatedAt: json['updatedAt']?.toString(),
    );
  }
}
